chainer/distributions/independent.py

```diff
@@ -0,0 +1,247 @@
+        elif reinterpreted_batch_ndims > len(distribution.batch_shape):
+            raise ValueError(
+                'reinterpreted_batch_ndims must be less than or equal to the '
+                'number of dimensions of `distribution.batch_shape`.')
+        self.__reinterpreted_batch_ndims = reinterpreted_batch_ndims
+
+        batch_ndim = \
+            len(self.distribution.batch_shape) - self.reinterpreted_batch_ndims
+        self.__batch_shape = distribution.batch_shape[:batch_ndim]
+        self.__event_shape = \
+            distribution.batch_shape[batch_ndim:] + distribution.event_shape
+
+    @property
+    def distribution(self):
+        return self.__distribution
+
+    @property
+    def reinterpreted_batch_ndims(self):
+        return self.__reinterpreted_batch_ndims
+
+    @property
+    def batch_shape(self):
+        return self.__batch_shape
+
+    @property
+    def event_shape(self):
+        return self.__event_shape
+
+    @property
+    def covariance(self):
+        '''Returns the covariance of the distribution based on the original
+        i.i.d. distribution. By definition, the covariance of the new
+        distribution becomes block diagonal matrix. Let
+        :math:`\\Sigma_{\\mathbf{x}}` be the covariance matrix of the original
+        random variable :math:`\\mathbf{x} \\in \\mathbb{R}^d`, and
+        :math:`\\mathbf{x}^{(1)}, \\mathbf{x}^{(2)}, \\cdots \\mathbf{x}^{(m)}`
+        be the :math:`m` i.i.d. random variables, new covariance matrix
+        :math:`\\Sigma_{\\mathbf{y}}` of :math:`\\mathbf{y} =
+        [\\mathbf{x}^{(1)}, \\mathbf{x}^{(2)}, \\cdots, \\mathbf{x}^{(m)}] \\in
+        \\mathbb{R}^{md}` can be written as
+
+        .. math::
+            \\left[\\begin{array}{ccc}
+                    \\Sigma_{\\mathbf{x}^{1}} & & 0 \\\\
+                    & \\ddots & \\\\
+                    0 & & \\Sigma_{\\mathbf{x}^{m}}
+            \\end{array} \\right].
+
+        Note that this relationship holds only if the covariance matrix of the
+        original distribution is given analytically.
+        '''
+        num_repeat = functools.reduce(
+            operator.mul,
+            self.distribution.batch_shape[-self.reinterpreted_batch_ndims:], 1)
+        dim = functools.reduce(operator.mul, self.distribution.event_shape, 1)
+        cov = repeat.repeat(
+            reshape.reshape(
+                self.distribution.covariance,
+                ((self.batch_shape) + (1, num_repeat, dim, dim))),
+            num_repeat, axis=-4)
+        cov = reshape.reshape(
+            transpose.transpose(
+                cov, axes=(
+                    tuple(range(len(self.batch_shape))) + (-4, -2, -3, -1))),
+            self.batch_shape + (num_repeat * dim, num_repeat * dim))
+        block_indicator = self.xp.reshape(
+            self._block_indicator,
+            tuple([1] * len(self.batch_shape)) + self._block_indicator.shape)
+        return cov * block_indicator
```

Comment:
To make a block diagonal matrix, the backward of `F.diagonal` (or the backward of unary `F.einsum`) could be used.

https://github.com/chainer/chainer/blob/c8c9ee0d5cdad4726f01e6a354f775c555f74355/tests/chainer_tests/functions_tests/math_tests/test_einsum.py#L232-L254